### PR TITLE
Modification for main modules compatibility

### LINKED ISSUE #816: Update GANDLF functionalities for compatibility with GAN pipelines
### Is your feature request related to a problem? Please describe.
To make new GAN utilities work, we need to slightly modify some of GANDLF utilities to also consider the generation problem.

### Describe the solution you'd like
Extension of logger, modification of modelio, modification of parameter processing and tensor files.

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #816

## Proposed Changes
- modification for proper scheduler and optimizer handling
- modification for proper onnx handling
- GAN logger instance
- adding model to the models registry

## Checklist

- [X] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [X] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [X] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [X] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [X] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

### CODE REVIEW

File: GANDLF/models/__init__.py
Comment (sarthakpati) [inline GANDLF/models/__init__.py]: Not sure if this is needed for **this** PR...?

File: GANDLF/models/__init__.py
Comment (sarthakpati) [inline GANDLF/models/__init__.py]: Not sure if this is needed for **this** PR...?

**Review by sarthakpati — COMMENTED** [review]
Most of it is fine. Just 1 question to address.

File: GANDLF/models/__init__.py
Comment (szmazurek) [reply GANDLF/models/__init__.py]: nah, you are right. Will move it into the dcgan branch

File: GANDLF/models/__init__.py
Comment (szmazurek) [reply GANDLF/models/__init__.py]: as above

File: GANDLF/models/__init__.py
Comment (sarthakpati) [reply GANDLF/models/__init__.py]: Thanks!

**Comment by szmazurek** [comment]
> Most of it is fine. Just 1 question to address.

Waiting for tests to pass, them will merge.
EDIT tests will not pass as there is an error due to generic function modification I made (I think I mentioned that it will fail, I expected that). Can I merge and then adress that?

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [inline GANDLF/utils/tensor.py]: Run `black` through this again.
```suggestion
                # check if any of the special cases are present
                if (
                    case in _class
                ):
```

File: GANDLF/data/ImagesFromDataFrame.py
Comment (sarthakpati) [inline GANDLF/data/ImagesFromDataFrame.py]:
```suggestion
        resized_image: str,
        output_dir: str,
        subject_id: str,
        channel_str: str,
        loader_type: str,
        extension: str,
```

File: GANDLF/logger.py
Comment (sarthakpati) [inline GANDLF/logger.py]: Update comment style.

File: GANDLF/logger.py
Comment (sarthakpati) [inline GANDLF/logger.py]: Update comment style.

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [inline GANDLF/utils/tensor.py]:
```suggestion
        optimizer_1 (torch.optim): The optimizer for training. In case of GANs, this is the optimizer for the generator.
        optimizer_2 (torch.optim): The second optimizer, used only in case of GANs. This is the optimizer for the discriminator.
```

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [inline GANDLF/utils/tensor.py]: Wondering if it would be better to have a list or dict of optimizers instead of just 2...?

**Review by sarthakpati — CHANGES_REQUESTED** [review]
What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?

Also, minor changes.

File: GANDLF/utils/tensor.py
Comment (szmazurek) [reply GANDLF/utils/tensor.py]: It is one of possible solutions, this was the first thing I come up with. Should I add that to this pr or we add it in the future?

**Comment by szmazurek** [comment]
> What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> 
> Also, minor changes.

I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?

**Comment by sarthakpati** [comment]
> > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > Also, minor changes.
> 
> I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?

How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...

**Comment by szmazurek** [comment]
> > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > Also, minor changes.
> > 
> > 
> > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> 
> How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...

I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving

**Comment by sarthakpati** [comment]
> > > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > > Also, minor changes.
> > > 
> > > 
> > > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> > 
> > 
> > How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...
> 
> I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving

Okay - do you mind setting it to the default (i.e., 120)? It's just that if this is not a parameter that we are controlling on the black configuration for GaNDLF, these changes will come up for each contributor... 🤕

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [reply GANDLF/utils/tensor.py]: Do you mind adding it now so that are _kind of_ future proofed? 😆

**Comment by szmazurek** [comment]
> > > > > What's the version of `black` you are using, BTW? Is it different from what's in [`setup.py`](https://github.com/mlcommons/GaNDLF/blob/master/setup.py#L85)?
> > > > > Also, minor changes.
> > > > 
> > > > 
> > > > I have the same, but probably I have different line length settings - I am in the 79 length clan, I believe default is 120 line length? Should I change that back?
> > > 
> > > 
> > > How are you suggesting black to make that change? I am curious because I don't think GaNDLF is setting that in the linting configuration...
> > 
> > 
> > I am setting that in the vscode black plugin settings, I have it configured that vscode automatically formats the document upon saving
> 
> Okay - do you mind setting it to the default (i.e., 120)? It's just that if this is not a parameter that we are controlling on the black configuration for GaNDLF, these changes will come up for each contributor... 🤕

I do not - will do that! And actually, default turns out to be 88 chars/line, so this is what I am going to set

File: GANDLF/logger.py
Comment (szmazurek) [reply GANDLF/logger.py]: Done

File: GANDLF/logger.py
Comment (szmazurek) [reply GANDLF/logger.py]: Done

File: GANDLF/utils/tensor.py
Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Added that, will still need to modify it in one place for GANs but that after the merge

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [inline GANDLF/utils/tensor.py]: Is the lines for `black` 120 or 79 for this file?

**Review by sarthakpati — COMMENTED** [review]
Looks good. Just confirm what the number of characters are for black for these files (it seems like it is less than what it was before).

File: GANDLF/utils/tensor.py
Comment (szmazurek) [reply GANDLF/utils/tensor.py]: It was with the default, so 88, I mentioned that in the last comment in the previous conversation. Should I go with 120 instead? BTW - I re-blacked once again all files in this pr, pushing now

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Ha! Strange thing, I manually moved it back to the version with this code in single line black did not move it back to the previous format. Dunno why I had to do it manually

File: GANDLF/logger.py
Comment (sarthakpati) [inline GANDLF/logger.py]: What is this comment about? 😆

File: GANDLF/utils/tensor.py
Comment (sarthakpati) [inline GANDLF/utils/tensor.py]: Seems like the line length is <88 for this block.

**Review by sarthakpati — CHANGES_REQUESTED** [review]
I don't think the `black` line-length setting was set properly. Perhaps this should be done using a project-level setting.

**Comment by sarthakpati** [comment]
> I don't think the `black` line-length setting was set properly. Perhaps this should be done using a project-level setting.

[This PR](https://github.com/mlcommons/GaNDLF/pull/820) should set this up correctly.

File: GANDLF/logger.py
Comment (szmazurek) [reply GANDLF/logger.py]: I was wondering what this line does :P But I managed to understand why it is like that, I think I will remove it :P

Comment (szmazurek) [reply GANDLF/logger.py]: removed

File: GANDLF/utils/tensor.py
Comment (szmazurek) [reply GANDLF/utils/tensor.py]: It was like 1 character (last comma in the brackets) that was preventing that from being set to single-line version. Removed it, also I have installed black from main branch, will use this one now.

Comment (sarthakpati) [reply GANDLF/utils/tensor.py]: Oh okay. The change doesn't seem to have worked for all files, though. For example:

image

Can you confirm?

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Look -  when I ran the black from CMD, there was no difference from the file I have commited here. However, some lines  got changed when I passed the -C flag, which stands for "don't go to newline if there is a trailing comma" (you can see it with black --help). Regarding the line you mentioned, take a look at my editor - I have a ruler set for 88 chars, and this one is slightly over that:
image

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Pushed the files ran with black with the commands I mentioned, give them a look and let's see if it did the trick. If so, maybe the -C flag should go into main branch for gandlf in extension config?

Comment (sarthakpati) [reply GANDLF/utils/tensor.py]: > -C flag should go into main branch for gandlf in extension config

Do you know how to set this up on the config? If so, can you please put in the PR?

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Short answer - no idea, but will give it a shot

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Edit - there are cases online where people set that in toml. I did it and installed correctly, however vscode does not give a damn about that for some reason. Explicitly setting this in the vscode ext settings works, but from cmd I still need the -C flag. No idea how to get over that, I have separate branch ready if you still want it to be commited

Comment (sarthakpati) [reply GANDLF/utils/tensor.py]: Yes, let's put that in as a PR. Thanks!

Comment (szmazurek) [reply GANDLF/utils/tensor.py]: Ok done. Closing then, if nothing elese then I am merging

**Review by sarthakpati — APPROVED** [review]

**Pull request merged**
